libcloud/storage/base.py

```diff
@@ -622,200 +624,62 @@ def _upload_object(self, object_name, content_type, upload_func,
-        if calculate_hash:
-            data_hash = data_hash.hexdigest()
-
-        return True, data_hash, bytes_transferred
-
-    def _stream_data(self, response, iterator, chunked=False,
-                     calculate_hash=True, chunk_size=None, data=None):
-        """
-        Stream a data over an http connection.
-
-        :param response: RawResponse object.
-        :type response: :class:`RawResponse`
-
-        :param response: An object which implements an iterator interface
-                         or a File like object with read method.
-        :type iterator: :class:`object`
-
-        :param chunked: True if the chunked transfer encoding should be used
-                        (defaults to False).
-        :type chunked: ``bool``
-
-        :param calculate_hash: True to calculate hash of the transferred data.
-                               (defaults to True).
-        :type calculate_hash: ``bool``
-
-        :param chunk_size: Optional chunk size (defaults to ``CHUNK_SIZE``)
-        :type chunk_size: ``int``
-
-        :rtype: ``tuple``
-        :return: First item is a boolean indicator of success, second
-                 one is the uploaded data MD5 hash and the third one
-                 is the number of transferred bytes.
-        """
-
-        chunk_size = chunk_size or CHUNK_SIZE
-
-        data_hash = None
-        if calculate_hash:
-            data_hash = self._get_hash_function()
-
-        generator = libcloud.utils.files.read_in_chunks(iterator, chunk_size,
-                                                        fill_size=True)
-
-        bytes_transferred = 0
-        try:
-            chunk = next(generator)
-        except StopIteration:
-            # Special case when StopIteration is thrown on the first iteration
-            # create a 0-byte long object
-            chunk = ''
-            if chunked:
-                response.connection.connection.send(b('%X\r\n' %
-                                                      (len(chunk))))
-                response.connection.connection.send(chunk)
-                response.connection.connection.send(b('\r\n'))
-                response.connection.connection.send(b('0\r\n\r\n'))
-            else:
-                response.connection.connection.send(chunk)
-            return True, data_hash.hexdigest(), bytes_transferred
-
-        while len(chunk) > 0:
-            try:
-                if chunked:
-                    response.connection.connection.send(b('%X\r\n' %
-                                                          (len(chunk))))
-                    response.connection.connection.send(b(chunk))
-                    response.connection.connection.send(b('\r\n'))
+        if upload_func:
+            upload_func(**upload_func_kwargs)
+
+        return {'response': response,
+                'bytes_transferred': stream_length,
+                'data_hash': stream_hash}
+
+    def _hash_buffered_stream(self, stream, hasher, blocksize=65536):
```

Comment:
I wonder why we have two implementations floating around in this file. Why the special casing for `__exit__` and `__next__`? This is a code smell for me. I feel that if we would refactor this function a little bit, we could do with a lot less complexity.

Can you perhaps add type hinting to this method so that it is clear what this method expects?